### PR TITLE
 [FLINK-19473] Implement multi inputs sorting DataInput 

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

This PR implements sorting inputs for multi input operators/tasks. It is based on the work in #13521
First I sort all inputs individually and after that I do a sorting merge of all the inputs to make sure that operator sees all incoming records with the same key from all inputs.

## Brief change log
For a more thorough explanation see the commits messages. Some highlights:
* Some of the commits refactor the TwoInputStreamProcessor/MultiInputStreamProcessor etc. to reduce code duplication
* Some of the commits extract the instantiation logic out of the processor and moves it to the Task. It makes it a) easier to instantiate in tests b) we need to pass less arguments to the ctor from the Task
* One of the commits implements proper `AvailabilityProvider` handling logic, which did not work before.

## Verifying this change

Added tests in:
* MultiInputSortingDataInputsTest
* SortedDataInputITCase
* SortingBoundedInputITCase.java 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by dawidwys** [comment]
@flinkbot run azure

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java:23]: re
> [hotfix] Fix availability handling in TwoInputStreamOperator

0. Can you explain a bit more what's the actual issue (I don't see what is this commit fixing)?
1. Can you create a bug ticket in the jira about about this and change commit from hotfix to [FLINK-XYZ]?
2. Definitely this commit is missing some test. If you will struggle to reproduce (`StreamTwoInputProcessor` is tested on the task level in `StreamTaskSelectiveReadingTest` and `TwoInputStreamTaskTest`. Note that `TwoInputStreamTaskTestHarness` that is used there is in the process of being deprecated and:
  a) it would be just nice to implement the test based on `StreamTaskMailboxTestHarness`. So far it's mostly used for multiple input tasks tests, but it has been used in source tasks and one input tasks tests as well (for example `StreamTaskTest#testNotifyCheckpointOnClosedOperator`, so it should be easy to use for the two inputs as well.
  b) mailbox version of the test harness might be much much better at reproducing a problem that depends on timing as you can execute it step by step
3. it would be best actually to merge this commit in separate PR to verify the performance impact of this bug fix alone (to narrow down suspects in case of regression detected after merging)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/TwoInputSelectionHandler.java
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/TwoInputSelectionHandler.java:73]: nit: I would flip the method to `isInputAvailable(...)` and use it with `!` operator

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java:67]: I don't get this condition. Why first has to be unavailable, while second available?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java:123]: why do we need this? The currently processed input should be updated via the returned `InputStatus`, no need to check it. If the other input is unavailable, and both will become unavailable, we would do the full check via waiting on availability future and we update the inputs (via cheap `isAproximatelyAvailable()`) once future is completed.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java:173]: Again as explained above, `StreamTwoInputProcessor#updateAvailability()` should do the trick (% the starvation check in L181)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java:91]: (I remember someone complaining about unchecked warnings 🙈 and `Object`? 😈  )

**Review by pnowojski — CHANGES_REQUESTED** [review]
I really like most of the commits here. Well done @dawidwys with the refactoring/cleanups/deduplicating the code.

I think my biggest concerns are about the bug fix commit hidden in this PR.

Note, I didn't review the `MultiInputSortingDataInputs` code itself, I hope @aljoscha will be able to do the most work here. Please let me know if you would like me to take a look at something there after all.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java:67]: Hah. That's actually a mistake. I forgot to remove the negation after I negated the method itself...

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/TwoInputSelectionHandler.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/TwoInputSelectionHandler.java:73]: I started with what you said, but then it is only ever used negated, that's why I flipped. If you prefer `isInputAvailable` I can do change that.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:214]: Yes, you are correct.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/EndOfInputAware.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/EndOfInputAware.java:40]: Ok, I will reuse the `BoundedMultiInput`, even though I am only like 90% convinced, because I am not a fan of reusing a single interface for multiple purposes. In this case one can argue though `OperatorChain` is an `Operator` and thus e.g. javadoc still makes sense.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInputs.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInputs.java:60]: It is. It is used in the `StreamTwoInputProcessorFactory/StreamMultipleInputProcessorFactory` to wrap inputs into sorted inputs.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java:91]: Yes, I must admit I was a bit harsh on this one :sweat_smile:

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java:23]: I had another look at the issue and now that I understand a bit better the assumptions on `getAvailability` I think the current code is ok % maybe we should change `StreamTwoInputProcessor#L206` to:

```
if (status == InputStatus.MORE_AVAILABLE || (status != InputStatus.END_OF_INPUT && (input.isApproximatelyAvailable() || input.isAvailable()))) {
```
My problem was that I did not use the `AvailabilityProvider.AVAILABLE` at all. My code was something like:

```
class MyInput {
   CompletableFuture<?> availabilityMarker = ...

   void executeSomeAsyncOperation() {
      availabilityMarker = new CompletableFuture();
      executeAsync(this::asyncOperation);
   }

   void asyncOperation() {
     ....
     availabilityMarker.complete(null);
   }

   CompletableFuture<?> getAvailability() {
      return availabilityProvider;
   }
}
```

This won't work because the default implementation of `isApproximatelyAvailable()` is never true. I would have to change:

```
   void asyncOperation() {
     ....
     CompletableFuture<?> tmp = availabilityMarker;
     availabilityMarker = AvailabilityProvider.AVAILABLE;
     tmp.complete(null);
   }
```

This is an implicit, not obvious contract that I was unaware of. The change I suggested for `StreamTwoInputProcessor#L206` would probably make it work for my old version, but I think it is not strictly necessary. On the other hand some additional documentation or a walkthrough example could be helpful here, especially as we are exposing this feature in user facing APIs.

I will remove this commit. Sorry for the confusion.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/sort/SortedInputITCase.java
Comment (dawidwys) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/sort/SortedInputITCase.java:64]: Good idea.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java:123]: The commit is unnecessary. I did not fully understand the availability handling.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java:173]: The commit is unnecessary. I did not fully understand the availability handling.

**Comment by aljoscha** [comment]
Just one comment regarding commits tags. I mostly use `[minor]` when it's minor refactoring changes because `[hotfix]` to me implies a fix for sth that was broken. Lately I have also seen Stephan do commits with just `[refactor] ...`, that's also an option.

**Review by aljoscha — COMMENTED** [review]
Overall the changes look very good! I had one comment before about commit tags and I had some inline comments about exception messages and some possible code changes.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInputs.java
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInputs.java]: Ok, usually I would suggest to squash such commits (to avoid having a dead code), but that's a minor issue, while indeed otherwise those two squashed commits would be quite big, so I'm fine with leaving it as it is.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java:23]: Ok, I see, and I agree that your change would increase usability of this code, protecting against those kind of bugs. However as you proposed it would currently add an extra `volatile` check. I think this could be addressed, by combining this check with anti-starvation check. However on the other hand...

It looks like I have implemented it differently in the `StreamMultipleInputProcessor`, where I do not see such check at all, and the processor completely relies on the `InputStatus`, which all in all I think is easier way of handling this. `isAvailable()` is checked only after the first `processInput` call after being woken up:
```
private int selectNextReadingInputIndex() {
	if (!inputSelectionHandler.isAnyInputAvailable()) {
		fullCheckAndSetAvailable();
	}
	...
}
``` 

Regarding the documentation. `AvailabilityProvider` already has quite a bit of java docs, but maybe you could help updating it to avoid such confusion in the future?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/TwoInputSelectionHandler.java
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/TwoInputSelectionHandler.java]: +0.1 for `isInputAvailable` from my side. `isInputUnavailable` sounds a bit strange :)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:68]: It emits the maximum watermark it has seen, MAX_WATERMARK could be interpreted as `Long.MAX_VALUE`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:129]: Indentation seems off here. (Another point for having an automatic formatting tool... 😅)

**Review by aljoscha — APPROVED** [review]
This looks very good! +1 to merge once CI is green.

I had two nitpicks about indentation and Javadocs.

**Comment by dawidwys** [comment]
@pnowojski I'd like to double check with you, are you fine with the current state of the PR?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/TwoInputSelectionHandler.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/TwoInputSelectionHandler.java]: I removed the commit, so I am not adding any of the two methods ;)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java:74]: Why do we need this commit? The default implementations should be doing the same thing.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java:53]: nit: it would be nice to keep the remaining processors in sync with this change?

**Review by pnowojski — COMMENTED** [review]
LGTM % a couple of nits

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java:74]: It is the same if the input does not override default implementations.

Mine does.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamOneInputProcessor.java:53]: `StreamMultipleInputProcessor` needs the `OperatorChain` for extracting chained sources. I updated it in the `StreamTwoInputProcessor` though.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:301]: Why are you implementing `isAvailable` and `isApproximatelyAvailable` independently? There should be no performance benefit of the default implementation `getAvailableFuture() == AVAILABLE`, and I would be afraid that it can create opportunities for those methods to drift out of sync.

So far in other places we were able to keep relatively simple contract, with controlling the availability via marking `AvailabilityHelper` available or not. Maybe let's sync off line on Monday?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java]: nitty nit: move `inputSelectionHandler.nextSelection();` outside of the if/else?

**Review by pnowojski — APPROVED** [review]
one optional nit besides, LGTM (assuming azure will be green).

Thanks for the changes @dawidwys.

**Pull request closed without merging**
